libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -97,321 +97,412 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
 void AC_AttitudeControl::input_rate_bf_roll_pitch_yaw(float roll_rate_bf_cds, float pitch_rate_bf_cds, float yaw_rate_bf_cds)
 {
     // Convert from centidegrees on public interface to radians
-    float roll_rate_bf_rads = radians(roll_rate_bf_cds*0.01f);
-    float pitch_rate_bf_rads = radians(pitch_rate_bf_cds*0.01f);
-    float yaw_rate_bf_rads = radians(yaw_rate_bf_cds*0.01f);
-
-    // Compute acceleration-limited body-frame roll rate
-    if (get_accel_roll_max_radss() > 0.0f) {
-        float rate_change_limit_rads = get_accel_roll_max_radss() * _dt;
-        _att_target_ang_vel_rads.x += constrain_float(roll_rate_bf_rads - _att_target_ang_vel_rads.x, -rate_change_limit_rads, rate_change_limit_rads);
-    } else {
-        _att_target_ang_vel_rads.x = roll_rate_bf_rads;
-    }
+    float roll_rate_rads = radians(roll_rate_bf_cds*0.01f);
+    float pitch_rate_rads = radians(pitch_rate_bf_cds*0.01f);
+    float yaw_rate_rads = radians(yaw_rate_bf_cds*0.01f);
 
-    // Compute acceleration-limited body-frame pitch rate
-    if (get_accel_pitch_max_radss() > 0.0f) {
-        float rate_change_limit_rads = get_accel_pitch_max_radss() * _dt;
-        _att_target_ang_vel_rads.y += constrain_float(pitch_rate_bf_rads - _att_target_ang_vel_rads.y, -rate_change_limit_rads, rate_change_limit_rads);
-    } else {
-        _att_target_ang_vel_rads.y = pitch_rate_bf_rads;
-    }
+    // calculate the attitude target euler angles
+    _attitude_target_quat.to_euler(_attitude_target_euler_angle.x, _attitude_target_euler_angle.y, _attitude_target_euler_angle.z);
 
-    // Compute acceleration-limited body-frame yaw rate
-    if (get_accel_yaw_max_radss() > 0.0f) {
-        float rate_change_limit_rads = get_accel_yaw_max_radss() * _dt;
-        _att_target_ang_vel_rads.z += constrain_float(yaw_rate_bf_rads - _att_target_ang_vel_rads.z, -rate_change_limit_rads, rate_change_limit_rads);
+    if (_rate_bf_ff_enabled) {
+        // Compute acceleration-limited euler rates
+        // When acceleration limiting is enabled, the input shaper constrains angular acceleration about the axis, slewing
+        // the output rate towards the input rate.
+        _attitude_target_ang_vel.x = input_shaping_ang_vel(_attitude_target_ang_vel.x, roll_rate_rads, get_accel_roll_max_radss());
+        _attitude_target_ang_vel.y = input_shaping_ang_vel(_attitude_target_ang_vel.y, pitch_rate_rads, get_accel_pitch_max_radss());
+        _attitude_target_ang_vel.z = input_shaping_ang_vel(_attitude_target_ang_vel.z, yaw_rate_rads, get_accel_yaw_max_radss());
     } else {
-        _att_target_ang_vel_rads.z = yaw_rate_bf_rads;
+        // When axis acceleration limiting is disabled, the attitude target is simply rotated using the input rate and the input rate
+        // is fed forward into the rate controller.
+        _attitude_target_ang_vel.x = roll_rate_rads;
+        _attitude_target_ang_vel.y = pitch_rate_rads;
+        _attitude_target_ang_vel.z = yaw_rate_rads;
     }
 
```

Comment:
Ok,
I changed the code to zero any rates if feedforward is disabled. It now calculates the new desired attitude and passes that into the attitude_controller_run_quat.